### DISCUSSION TITLE
Contributing CDMC Compliant Services into Compliant Financial Infrastructure

### DISCUSSION THREAD

**Discussion by mcleo-d** [finos/compliant-financial-infrastructure#197] [FINOS and CDMC Roadmap]
## Description

The original workflow communicated when the CDMC joined forces with Compliant Financial Infrastructure and Open Complicate for CDMC is creating a lot of confusion as it communicates a waterfall contribution mechanism that originates from the CDMC.

The original workflow can be found on GitHub Issue https://github.com/finos/compliant-financial-infrastructure/issues/156.

This discussion has been created to illustrate a new contribution workflow that's more iterative and where Compliant Financial Infrastructure contributions can be made from across the FINOS open source community.

The workflow below should demonstrate a non-blocking, Apache 2, iterative mechanism with feedback loops and a continuous cycle of contribution and communication. Please highlight if this is not the case!

![Contributing-CDMC-Compliant-Services](https://user-images.githubusercontent.com/6029572/157042229-8f002572-b9d8-45a7-bd73-b0db73b35f1d.png)

It would be great to get CFI and OC-CDMC agreement on the workflow on **Friday 11th January**, so please use this discussion to evolve the mechanism before the next project meeting.

Thanks for your continuous input and support.

**Comment by mcleo-d** [comment]
To remove any ambiguity, I have created the following walkthrough video of the proposed Compliant Financial Infrastructure contribution workflow below.

https://user-images.githubusercontent.com/6029572/157091339-1f3da590-e35a-486d-a5e2-462518d65d70.mov

**Comment by thinkl33t** [comment]
This looks like an excellent workflow for closing the gap between the CDMC document and our testing within FinOS, great work all involved!

I guess practically to implement it, the first thing we'd need to do would be to work over the existing Service Accelerators in CFI, figure out mappings between the specifications in them and the controls in the CDMC document, and fill in any gaps.

**Reply by mcleo-d** [reply]
Thanks @thinkl33t - I’d like to add this to Friday’s agenda to help overcome the bottleneck we’re facing of that works for you?

@eddie-knight also has a suggestion for how repos should be organised for Terraform and Probr, which is coming next.

**Comment by eddie-knight** [comment]
I've drafted a little diagram with a proposed reorganization, based on the following considerations:

   1. Considering likely customer use cases, we should create IaC that is registry-ready. For Terraform Enterprise, this means creating one repo per module.
   2. Probr test packs are intended to be independently versioned binaries. The current boilerplate approach to creating a test pack involves creating a repository and using the Github "releases" feature.
   3. Service Accelerator Templates (SAT) are individual files that can easily share a repository with CFI documentation
   4. Git modules are a very clear (though occasionally clunky) strategy for linking repositories. The CFI repo can be used to topically organize the SAT and links to the IaC and applicable test packs.

![CFI Repo Organization](https://user-images.githubusercontent.com/21176439/157146521-9183b694-8c96-41b3-bae6-9740fbfefb74.png)

**Reply by mcleo-d** [reply]
Hi @eddie-knight,

Thanks for your input into how the project structure will need to shift as we start developing Probr tests and Terraform Enterprise modules.

It would be great to know whether technology actually drives how the repos should be structured and whether there is much room for opinion and debate?  

I personally like the separation of concerns, but is it the case that technology is driving this change, and we should be educating the group on how to move forward rather than taking too much feedback or asking for alternatives? 🤔 

James.
